flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -290,17 +290,33 @@ private void registerCallback() {
     @Override
     public void write(InputT element, Context context) throws IOException, InterruptedException {
         while (bufferedRequestEntries.size() >= maxBufferedRequests) {
-            mailboxExecutor.tryYield();
+            flush();
```

Comment:
8. Note, this is the only time in the write method that we call flush without checking whether it will block or not. This is precisely when our buffer is full. In this case, we have no choice but to do the only action that has a chance of reducing the number of elements in the buffer - which is to flush. If there are free in flight requests then flush will be non blocking and complete after firing off that in flight request. If there are NO free in flight requests, then this call will block until the number of in flight requests is less than the allowed number.

Therefore an important corollary appears: "calls to write will block if and only if the buffer is full and the number of in flight requests is equal to or greater than the maximum allowed".

Reply:
9. Additionally to align with @pnowojski 's example, in this sink, we:
a) start flushing if `sizeOfBuffer >= maxBatchSize` or `sizeOfBufferBytes >= maxBatchSizeInBytes`
b) block new flushes once `inFlightRequestsCount` >= `maxInFlightRequests`
 - unblock as soon as the condition does not hold true
c) block new writes if the buffer is full
 - unblock as soon as the condition does not hold true